frostsnap_core/src/device.rs

```diff
@@ -254,129 +243,181 @@ impl FrostSigner {
                     .public_key()
                     .to_xonly_bytes();
 
-                self.state = SignerState::FrostKey {
-                    key: FrostsnapKey {
+                self.action_state = Some(SignerState::KeyGenAck {
+                    key: FrostsnapSecretKey {
                         frost_key,
                         secret_share,
                         share_index: *my_index,
-                        aux_rand,
                     },
-                    awaiting_ack: true,
-                };
+                });
 
                 Ok(vec![DeviceSend::ToUser(DeviceToUserMessage::CheckKeyGen {
                     session_hash,
                 })])
             }
             (
-                SignerState::FrostKey {
-                    key,
-                    awaiting_ack: false,
-                },
+                None,
                 CoordinatorToDeviceMessage::RequestSign(SignRequest {
                     nonces,
                     sign_task,
                     key_id,
                 }),
             ) => {
-                let (my_nonces, my_nonce_index, _) = match nonces.get(&key.share_index) {
-                    Some(nonce) => nonce,
-                    None => return Ok(Vec::new()),
-                };
+                let key = self.keys.get(&key_id).ok_or_else(|| {
+                    Error::signer_invalid_message(
+                        &message,
+                        // we could instead send back a message saying we don't have this key but I
+                        // think this will never happen in practice unless we have a way for one
+                        // coordinator to delete a key from a device without the other coordinator
+                        // knowing.
+                        format!("device doesn't have key for {key_id}"),
+                    )
+                })?;
+
+                let key_id = key.key_id();
+                let my_nonces = nonces.get(&key.share_index).ok_or_else(|| {
+                    Error::signer_invalid_message(
+                        &message,
+                        "this device was asked to sign but no nonces
+                were provided",
+                    )
+                })?;
 
                 let n_signatures_requested = sign_task.sign_items().len();
-                if my_nonces.len() != n_signatures_requested {
-                    return Err(Error::signer_invalid_message(&message, format!( "Number of nonces ({}) was not the same as the number of signatures we were asked for {}", my_nonces.len(), n_signatures_requested)));
+                if my_nonces.nonces.len() != n_signatures_requested {
+                    return Err(Error::signer_invalid_message(&message, format!("Number of nonces ({}) was not the same as the number of signatures we were asked for {}", my_nonces.nonces.len(), n_signatures_requested)));
```

Comment:
I'm surprised this passes clippy fmt, but I also have had trouble formatting multiline strings in rust 😅

Reply:
cargo fmt ignores macros because it's nt sure how changing the contents will change the semantics I guess.